modules/fedimint-walletv2-server/src/lib.rs

```diff
@@ -0,0 +1,1147 @@
+
+        counts.get(num_peers.threshold() - 1).copied().unwrap_or(0)
+    }
+
+    pub async fn consensus_feerate(&self, dbtx: &mut DatabaseTransaction<'_>) -> Option<u64> {
+        let transaction_count = self.consensus_transaction_count(dbtx).await;
+
+        // Stacking 45 unconfirmed transactions on top of each other would require a
+        // feerate over 50msats << 44 satoshis per virtual byte.
+
+        assert!(transaction_count < 45);
+
+        // Due to the above assertion the following shifts cannot wrap.
+
+        self.consensus_feerate_median(dbtx)
+            .await
+            .map(|mut msats_per_vbyte| {
+                // Unconfirmed | Last Multiplier | Average Multiplier
+                // ------------|-----------------|-------------------
+                // 1           | 1.00            | 1.00
+                // 2           | 1.33            | 1.17
+                // 3           | 1.78            | 1.37
+                // 4           | 2.37            | 1.62
+                // 5           | 3.16            | 1.93
+                // 6           | 4.21            | 2.31
+                // 7           | 5.62            | 2.78
+                // 8           | 7.49            | 3.28
+
+                for _ in 0..transaction_count {
+                    msats_per_vbyte = msats_per_vbyte.saturating_mul(4).saturating_div(3);
+                }
+
+                // A stack of 25 unconfirmed transactions will have a minimum average feerate
+                // of (50msats << 25) / 25 / 1000 = 67.108  satoshis per virtual byte. This
+                // protects us against a otherwise catastrophic failure of feerate estimation.
+
+                msats_per_vbyte.max(50 << transaction_count)
+            })
+    }
+
+    pub async fn consensus_feerate_median(
+        &self,
+        dbtx: &mut DatabaseTransaction<'_>,
+    ) -> Option<u64> {
+        let num_peers = self.cfg.consensus.bitcoin_pks.to_num_peers();
+
+        let mut rates = dbtx
+            .find_by_prefix(&FeeRateVotePrefix)
+            .await
+            .filter_map(|entry| async move { entry.1 })
+            .collect::<Vec<u64>>()
+            .await;
+
+        assert!(rates.len() <= num_peers.total());
+
+        rates.sort_unstable();
+
+        assert!(rates.first() <= rates.last());
+
+        rates.get(num_peers.threshold() - 1).copied()
+    }
+
+    async fn consensus_transaction_count(&self, dbtx: &mut DatabaseTransaction<'_>) -> u32 {
+        let unsigned_count = dbtx
+            .find_by_prefix(&UnsignedTransactionPrefix)
+            .await
+            .count()
+            .await;
+
+        let pending_count = dbtx
+            .find_by_prefix(&PendingTransactionPrefix)
+            .await
+            .count()
+            .await;
+
+        // Since the total number of transactions is bound the following addition
+        // and cast cannot overflow or truncate.
+
+        (unsigned_count + pending_count) as u32
```

Comment:
```suggestion
        (unsigned_count + pending_count).try_into().expect("Since the total number of transactions is bound the addition cannot overflow or truncate")
```